### DISCUSSION TITLE
Mutiple entries which have its own hmr runtime cause hmr fail

### DISCUSSION THREAD

**Discussion by nyqykk** [webpack/webpack#17933] [Q&A]
# Bug report
One project use mutiple entries in some cases will build hmr runtime into both entries(can not avoid). This globalFuntion will be override by the last entry, cause hmr fail https://github.com/webpack/webpack/blob/main/lib/web/JsonpChunkLoadingRuntimeModule.js#L339

**What is the current behavior?**
hmr fail, strictly reload.

**If the current behavior is a bug, please provide the steps to reproduce.**
I rewrite hmr runtime can solve this problem. I just inherit the globalFuntion and use global variable to store the updateQueue
<img width="1220" alt="image" src="https://github.com/webpack/webpack/assets/65393845/ffc5dec8-5c42-4dfb-aef5-9e6e55228a68">
<img width="867" alt="image" src="https://github.com/webpack/webpack/assets/65393845/dbc8fcca-8828-4790-8bb1-587ec172d332">

**What is the expected behavior?**

**Other relevant information:**
webpack version:  5.89.0
Node.js version: 16.18.1
Operating System: macos
Additional tools: no

**Comment by alexander-akait** [comment]
Please create a reproducible test repository, thank you

**Reply by nyqykk** [reply]
i have provide a cloud ide. In the repo, i add a custom plugin to add a entry which will not be splitChunk. This case may usually appear in Micro-Frontend mode. You can imagine a Master App and a Child App, if Child App split chunk and some chunk are built into html as script tag. In this case if Master App load Child App's js file, some chunks will not be loaded forever, because it is build into html template. So Child App can not work.

In the repo, main entry and second entry have its own hmr runtime, and the second entry is loaded after main entry, so the global hmr function will be override by second entry.
<img width="1891" alt="image" src="https://github.com/webpack/webpack/assets/65393845/70bfeea9-8322-4098-ba9f-4f1571c1cfc2">

Step:
npm start -> refresh page -> modify second entry -> hmr success
npm start -> refresh page -> modify main entry -> hmr fail

https://stackblitz.com/edit/webpack-webpack-js-org-q8kgen?file=webpack.config.js,public%2Fsecond-entry.js,src%2Findex.js,public%2Findex.html,public%2Fadd-entry-plugin.js&terminal=build,start

**Reply by nyqykk** [reply]
sorry, is there any problems for this repo

**Reply by alexander-akait** [reply]
Sorry for delay, because you use two chunks on the page you need to set `runtime` and don't forget to unclude:

```js
 optimization: {
    runtimeChunk: {
      name: 'runtime',
    },
},
```

```html
<script defer src="runtime.js"></script>
<script defer src="main.js"></script>
<script defer src="second-entry.js"></script>
```

**Reply by nyqykk** [reply]
Yes, It is worked when this an independent application. if we set runtimeChunk as 'single', main entry and the second entry will use the same runtime.js as there runtimeChunk. It perform that both two entry call __webpack_require__.o firstly to ensure runtimeChunk are loaded and then call __webpack_require__ to get the entry exports.

But i mean in Micro Frontend mode, this is a child application, and main.js and the second entry are the entry of the child application. If runtimeChunk are split into runtime.js file, when we load main.js or the second entry js file, runtime.js will not be loaded and then the application can not start up because both two entry miss some runtime chunk. 
This perform that both two entry also call __webpack_require__.o to ensure runtimeChunk are loaded, but it will not loaded forever because we split it into other runtime.js file, so the application can not work.

**Reply by alexander-akait** [reply]
In this case please use multi compiler mode - https://github.com/webpack/webpack/blob/main/examples/multi-compiler/README.md and set different name for the each compilation, also we have https://webpack.js.org/configuration/other-options/#dependencies to set the order.

**Comment by nyqykk** [comment]
ok, that's a solution for me, thanks for help~
